internal/consensus/block_executor.go

```diff
@@ -0,0 +1,120 @@
+package consensus
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	cstypes "github.com/tendermint/tendermint/internal/consensus/types"
+	sm "github.com/tendermint/tendermint/internal/state"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/types"
+)
+
+type blockExecutor struct {
+	logger             log.Logger
+	privValidator      privValidator
+	blockExec          sm.Executor
+	proposedAppVersion uint64
+}
+
+// Create the next block to propose and return it. Returns nil block upon error.
+//
+// We really only need to return the parts, but the block is returned for
+// convenience so we can log the proposal block.
+//
+// NOTE: keep it side-effect free for clarity.
+// CONTRACT: cs.privValidator is not nil.
+func (c *blockExecutor) create(ctx context.Context, stateData *StateData, round int32) (*types.Block, error) {
+	if c.privValidator.IsZero() {
+		return nil, errors.New("entered createProposalBlock with privValidator being nil")
+	}
+
+	// TODO(sergio): wouldn't it be easier if CreateProposalBlock accepted cs.LastCommit directly?
+	var commit *types.Commit
+	switch {
+	case stateData.Height == stateData.state.InitialHeight:
+		// We're creating a proposal for the first block.
+		// The commit is empty, but not nil.
+		commit = types.NewCommit(0, 0, types.BlockID{}, nil)
+	case stateData.LastCommit != nil:
+		// Make the commit from LastPrecommits
+		commit = stateData.LastCommit
+
+	default: // This shouldn't happen.
+		c.logger.Error("propose step; cannot propose anything without commit for the previous block")
+		return nil, nil
+	}
+
+	proposerProTxHash := c.privValidator.ProTxHash
+
+	ret, uncommittedState, err := c.blockExec.CreateProposalBlock(ctx, stateData.Height, round, stateData.state, commit, proposerProTxHash, c.proposedAppVersion)
+	if err != nil {
+		panic(err)
+	}
+	stateData.RoundState.CurrentRoundState = uncommittedState
+	return ret, nil
+}
+
+func (c *blockExecutor) process(ctx context.Context, stateData *StateData, round int32) error {
```

Comment:
I renamed (simplified) `ensureProcessProposal` to `process`, because `ensuring` is the default behavior of this process.
I don't mind to rename the method name back, 
would you like to rename it?

Reply:
renamed on `ensureProcess`